File: 112.path-sum.py
#
# @lc app=leetcode id=112 lang=python3
#
# [112] Path Sum
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if root is None:
            return False
        if not root.left and not root.right and root.val == sum:
            return True
        return self.hasPathSum(root.left,sum-root.val) or self.hasPathSum(root.right,sum-root.val)

        
# @lc code=end

File: golang/112.path-sum.go
/*
 * @lc app=leetcode.cn id=112 lang=golang
 *
 * [112] Path Sum
 */
package main

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, target int) bool {
	return hasPathSumHelper(root, target, 0)
}

func hasPathSumHelper(root *TreeNode, target int, cur int) bool {
	if root == nil {
		return false
	}
	if root.Left == nil && root.Right == nil {
		if cur+root.Val == target {
			return true
		}
		return false
	}
	left := hasPathSumHelper(root.Left, target, cur+root.Val)
	right := hasPathSumHelper(root.Right, target, cur+root.Val)
	return left || right
}

// @lc code=end
